Tensor changes for 2 dataloaders and 2 augmentations

fiona-gladwin opened the pull request:
- Add tensor changes for rocalJpegFileSource and rocalJpegFileSourceSingleShard
- Modify create_image and create_loader_output_image in master_graph
- Add tensor changes for brightness and crop mirror normalize augmentations
- Modify the node_brightness and node_crop_mirror_normalize accordingly.
- Add support to pass input and output layouts to files in AMD OpenVX extensions
- Add support to pass ROI as pinned memory from rocAL.
- Introduce copy constructor and destructor for rocalTensorInfo class
- Modify set_layout API to change dimensions based on output layout
- Add INT8 and UINT32 support for vx_tensors

fiona-gladwin committed:
Add support to return rocALTensor for file source dataloader

Add changes in master graph to create tensor and loader tensor API

fiona-gladwin committed:
Add API to compute desc ptr values from dims

fiona-gladwin committed:
Brightness related changes

Add changes to brightness API
Add support to pass in and out layouts

fiona-gladwin committed:
CMN related changes

Add support to pass in and out layouts

fiona-gladwin committed:
Minor changes

Add tensor layout and tensor data type API

fiona-gladwin committed:
Change layout and roi type scalar to INT

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Change API in tensor files to modify dims based on layout

fiona-gladwin committed:
Minor code changes

fiona-gladwin committed:
Code clean up

fiona-gladwin committed:
Remove depth and start_z param from CMN

fiona-gladwin committed:
Remove depth and start_z param from CMN

fiona-gladwin committed:
Merge branch 'PR_tensor3' of https://github.com/fiona-gladwin/MIVisionX into PR_tensor3

fiona-gladwin committed:
Move ROI updation to node_crop_mirror_normalize

Update the ROI with the crop coordinates and pass to vx_rpp
Remove crop related params from CropMirrorNormalize.cpp

fiona-gladwin committed:
Modify mean and std dev args in CMN

Add support to pass mean and std dev values for RGB or Grayscale images

fiona-gladwin committed:
Modify set tensor layout API

Make changes to set max shape only once

fiona-gladwin committed:
Modify arguments arrangement

Brightness and CMN

fiona-gladwin committed:
Modify code format for curly braces

fiona-gladwin committed:
Minor changes

Rename arguments
Add comments

fiona-gladwin committed:
Add INT8 data type support

fiona-gladwin committed:
Minor code changes in vx_rpp files

fiona-gladwin committed:
Minor code changes in vx_rpp files

fiona-gladwin committed:
Merge branch 'PR_tensor3' of https://github.com/fiona-gladwin/MIVisionX into PR_tensor3

fiona-gladwin committed:
Add multiplier and offset change

fiona-gladwin committed:
Merge branch 'PR_tensor3' of https://github.com/fiona-gladwin/MIVisionX into PR_tensor3

fiona-gladwin committed:
Add INT8 and UINT32 support for vx tensor

fiona-gladwin committed:
Use vx_tensors for src and dst in node

Create vs tensors with the ROI buffer

fiona-gladwin committed:
Add support to allocate ROI as HOST or Device buffer

Add copy constructor and destructor to tensorInfo class

fiona-gladwin committed:
Add support to pass ROI as vx_tensor in OpenVX ext

Modify Brightness and CMN augmentations

fiona-gladwin committed:
Modify the SRC ROI for CMN Node

fiona-gladwin committed:
Minor change

Remove update array call while creating array in paramter_crop

fiona-gladwin committed:
Add changes to node image loader

fiona-gladwin committed:
Minor code changes

fiona-gladwin committed:
Minor fix - TensorInfo destructor

fiona-gladwin committed:
Minor change

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h:
```diff
@@ -30,6 +30,7 @@ THE SOFTWARE.
 #endif
 
 #include "internal_rpp.h"
+#define NUM_OF_DIMS 5
```
@fiona-gladwin  can you rename this to rpp_max_tensor_dims or similar? We already have another tensor dimensions set to max 6 [here](https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/blob/master/amd_openvx/openvx/ago/ago_internal.h#L107). Would want to avoid any confusion

rrawther replied:
Yes.

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:
```diff
@@ -2033,3 +2030,70 @@ vx_status releaseGraphHandle(vx_node node, RPPCommonHandle *handle)
+            descPtr->strides.nStride = descPtr->c * descPtr->w * descPtr->h;
+            descPtr->strides.hStride = descPtr->c * descPtr->w;
+            descPtr->strides.wStride = descPtr->c;
+            descPtr->strides.cStride = 1;
```
@fiona-gladwin shouldn't this stride be sizeof(data_type) and similarly the rest multiplied with the same value?
for ex: with vx_tensor of type float32, stride[0] = sizeof(float32), stride[1] = input_dims[0]*sizeof(float32) etc

fiona-gladwin replied:
Currently we are modifying the stride value in rocAL to match with stride calculations in RPP.

rrawther commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -128,9 +128,9 @@ extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_WarpPerspectivebatchP
 extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_Resizetensor(vx_graph graph,vx_image pSrc,vx_array srcImgWidth,vx_array srcImgHeight,vx_image pDst,vx_array dstImgWidth,vx_array dstImgHeight,vx_uint32 nbatchSize);
 
-extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_Brightness(vx_graph graph, vx_tensor pSrc, vx_array srcROI, vx_tensor pDst, vx_array alpha, vx_array beta, vx_scalar layout, vx_scalar roiType, vx_uint32 nbatchSize);
+extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_Brightness(vx_graph graph, vx_tensor pSrc, vx_tensor srcROI, vx_tensor pDst, vx_array alpha, vx_array beta, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType, vx_uint32 nbatchSize);
```
I think it is better to pass alpha and beta values to rpp through a single tensor pointer instead of 2 arrays

rrawther commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -128,9 +128,9 @@ extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_WarpPerspectivebatchP
+extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_Brightness(vx_graph graph, vx_tensor pSrc, vx_tensor srcROI, vx_tensor pDst, vx_array alpha, vx_array beta, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType, vx_uint32 nbatchSize);
 extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_Copy(vx_graph graph, vx_tensor pSrc, vx_tensor pDst);
-extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_CropMirrorNormalize(vx_graph graph, vx_tensor pSrc, vx_array srcROI, vx_tensor pDst, vx_array dstROI, vx_array crop_w, vx_array crop_h, vx_array x1, vx_array y1, vx_array mean, vx_array std_dev, vx_array flip, vx_scalar layout, vx_scalar roiType, vx_uint32 nbatchSize);
+extern  "C" SHARED_PUBLIC vx_node VX_API_CALL vxExtrppNode_CropMirrorNormalize(vx_graph graph, vx_tensor pSrc, vx_tensor srcROI, vx_tensor pDst, vx_array mean, vx_array std_dev, vx_array flip, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType, vx_uint32 nbatchSize);
```
do we need to pass batchSize as parameter? Can't the node query tensor for for the batchsize? It is assuming roiType to be same for both source and destination. Is that the case always?

fiona-gladwin replied:
In this node we are passing only srcROI hence only one roiType variable is passed which denotes type of the source ROI

fiona-gladwin replied:
Removed batchSize parameter

rrawther commented on amd_openvx_extensions/amd_rpp/source/Brightness.cpp:
```diff
@@ -21,100 +21,102 @@ THE SOFTWARE.
+#endif
     }
+    data->roiPtr = (RpptROI *)data->roiTensorPtr;
+    if((data->inputLayout == 2 || data->inputLayout == 3)) { // For NFCHW and NFHWC formats
```
Can't we pass the correct data pointer directly to Rpp instead of converting from one format to other? In that case only, we can use pinned memory in rocAL

fiona-gladwin replied:
This conversion happens for NFHWC/NFCHW formats currently. We will fix this in future PR

rrawther commented on amd_openvx_extensions/amd_rpp/source/Brightness.cpp:
```diff
@@ -152,209 +152,85 @@ static vx_status VX_CALLBACK validateBrightness(vx_node node, const vx_reference
 #if ENABLE_HIP
-    if (data->device_type == AGO_TARGET_AFFINITY_GPU)
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU)
         rppCreateWithStreamAndBatchSize(&data->rppHandle, data->handle.hipstream, data->nbatchSize);
```
as we discussed before, we should use a single handle for the entire pipeline instead of creating one for each node

fiona-gladwin replied:
Modified the codes to create a single handle for all RPP nodes. Added the changes in Brightness and CropMirrorNormalize nodes.

rrawther commented on amd_openvx_extensions/amd_rpp/source/CropMirrorNormalize.cpp:
```diff
@@ -21,127 +21,116 @@ THE SOFTWARE.
-    vx_enum out_tensor_type;
+    Rpp32s inputLayout;
+    Rpp32s outputLayout;
+    size_t inputTensorDims[NUM_OF_DIMS];
```
layout, dims and type has to be part of tensor descriptor

fiona-gladwin replied:
The tensor descriptor in RPP stores the dims in a particular structure format hence we are storing dims in the array and converting to the RPP struct format for now, Once change is made in RPP to accept dims array we will make changes accordingly in OpenVX extensions

rrawther commented on amd_openvx_extensions/amd_rpp/source/CropMirrorNormalize.cpp:
```diff
@@ -21,127 +21,116 @@ THE SOFTWARE.
+#endif
+    }
+    data->roiPtr = (RpptROI *)data->roiTensorPtr;
+    if(data->inputLayout == 2 || data->inputLayout == 3) { // For NFCHW and NFHWC formats
```
refer to previous comment regarding this

rrawther commented on amd_openvx_extensions/amd_rpp/source/CropMirrorNormalize.cpp:
```diff
@@ -179,220 +166,89 @@ static vx_status VX_CALLBACK validateCropMirrorNormalize(vx_node node, const vx_
     refreshCropMirrorNormalize(node, parameters, num, data);
 #if ENABLE_HIP
-    if (data->device_type == AGO_TARGET_AFFINITY_GPU)
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU)
```
refer to previous comment

rrawther commented on amd_openvx_extensions/amd_rpp/source/CropMirrorNormalize.cpp:
```diff
@@ -179,220 +166,89 @@ static vx_status VX_CALLBACK validateCropMirrorNormalize(vx_node node, const vx_
-    data->mean = (vx_float32 *)malloc(sizeof(vx_float32) * data->src_desc_ptr->n * 3);
-    data->std_dev = (vx_float32 *)malloc(sizeof(vx_float32) * data->src_desc_ptr->n * 3);
-    data->mirror = (vx_uint32 *)malloc(sizeof(vx_uint32) * data->src_desc_ptr->n);
+    data->mean = (vx_float32 *)malloc(sizeof(vx_float32) * data->srcDescPtr->n * data->srcDescPtr->c);
```
better to pass mean, std and mirror as a single tensor using pinned host mem

rrawther commented on amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:
```diff
@@ -1864,11 +1864,12 @@ VX_API_ENTRY vx_node VX_API_CALL vxExtrppNode_Brightness(vx_graph graph, vx_tens
+            (vx_reference)inputLayout,
+            (vx_reference)outputLayout,
             (vx_reference)roiType,
             (vx_reference)NBATCHSIZE,
```
please don't use capitals for parameters and infer batchsize from tensor instead of passing as a separate parameter

fiona-gladwin replied:
We have used capital for the BATCHSIZE and DEVTYPE variables alone because the vx_scalar is created in this function

rrawther commented on rocAL/rocAL/include/loaders/image/node_image_loader_single_shard.h:
```diff
@@ -29,9 +29,9 @@ class ImageLoaderSingleShardNode : public Node
 public:
 #if ENABLE_HIP
-    ImageLoaderSingleShardNode(Image *output, DeviceResourcesHip device_resources);
+    ImageLoaderSingleShardNode(rocalTensor *output, DeviceResourcesHip device_resources);
```
this need to be updated with tot changes which separates HIP, OpenCL and CPU

rrawther commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -71,38 +71,69 @@ class rocalTensorInfo {
+        if(_layout != layout && _layout != RocalTensorlayout::NONE) {
+            std::vector<size_t> new_dims(_num_of_dims, 0);
+            get_modified_dims_from_layout(_layout, layout, new_dims);
+            _dims = new_dims;
```
I am not sure why the dims needs to be modified?

fiona-gladwin replied:
There are instances where the input tensor and output tensor may have different layouts. When we are doing a copy of the input tensor info to the output info in the rocal_api_augmentations.cpp we may need to modify the dims to convert from one layout to another, i.e NCHW to NHWC and vice versa

rrawther commented on rocAL/rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -70,6 +70,45 @@ THE SOFTWARE.
 
+void get_rocal_tensor_layout(RocalTensorLayout &tensor_layout, RocalTensorlayout &op_tensor_layout)
+{
+    switch(tensor_layout)
```
What is the purpose of this api. We are just copying one tensor_layout to other?

rrawther commented on rocAL/rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -70,6 +70,45 @@ THE SOFTWARE.
+    }
+}
+
+void get_rocal_tensor_data_type(RocalTensorOutputType &tensor_output_type, RocalTensorDataType &tensor_data_type)
```
same here. Why do we need this function since it is just doing a copy

rrawther commented on rocAL/rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -605,33 +644,36 @@ rocalResize(
+    try {
+        RocalTensorlayout op_tensorLayout;
+        RocalTensorDataType op_tensorDataType;
+        get_rocal_tensor_layout(rocal_tensor_output_layout, op_tensorLayout);
```
if the output layout and type is the same as input, please get rid of the unnecessary functions and use the input layout and type directly

fiona-gladwin replied:
This has been removed in recent commits

rrawther commented on rocAL/rocAL/source/augmentations/color_augmentations/node_brightness.cpp:
```diff
@@ -25,43 +25,43 @@ THE SOFTWARE.
+    vx_scalar in_layout_vx = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &input_layout);
+    vx_scalar out_layout_vx = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &output_layout);
+    vx_scalar roi_type_vx = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &roi_type);
+    _node = vxExtrppNode_Brightness(_graph->get(), _inputs[0]->handle(), _src_tensor_roi, _outputs[0]->handle(), _alpha.default_array(), _beta.default_array(), in_layout_vx, out_layout_vx, roi_type_vx, _batch_size);
```
alpha and beta needs to be passed through a single tensor instead of 2 arrays and remove batch_size

rrawther commented on rocAL/rocAL/source/augmentations/geometry_augmentations/node_crop_mirror_normalize.cpp:
```diff
@@ -26,59 +26,91 @@ THE SOFTWARE.
-         _mean_vx[i] = _mean;
-         _std_dev_vx[i] = _std_dev;
+
+    std::vector<float> mean_vec, std_dev_vec;
```
looks like mean and std are converted to multiplier and offset. Please rename the variables to reflect that. Also we need to pass it using a single tensor if multiplier and offset are different for each image in tensor.

rrawther commented on rocAL/rocAL/source/parameters/parameter_crop.cpp:
```diff
@@ -76,7 +76,6 @@ void CropParam::create_array(std::shared_ptr<Graph> graph)
     vxAddArrayItems(croph_arr,batch_size, croph_arr_val.data(), sizeof(vx_uint32));
     vxAddArrayItems(x2_arr,batch_size, x2_arr_val.data(), sizeof(vx_uint32));
     vxAddArrayItems(y2_arr,batch_size, y2_arr_val.data(), sizeof(vx_uint32));
-    update_array();
```
why is this removed?

fiona-gladwin replied:
In this create_array API, the crop related arrays are created and updated with the crop values, but when it is invoked by the create_node API in the nodes the ROI value is still not updated, hence the crop arrays are not updated correctly. So this update_array call is not required here

rrawther commented on rocAL/rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -296,36 +296,37 @@ MasterGraph::build()
-            THROW("Cannot create the image from handle")
+rocalTensor *
+MasterGraph::create_tensor(const rocalTensorInfo &info, bool is_output) {
+    auto *new_tensor = new rocalTensor(info);
```
please use output instead of new_tensor

fiona-gladwin replied:
Renamed 
new_tensor -> output
output -> output_ext (since this tensor will be returned and accessed by the user)

rrawther commented on rocAL/rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -296,36 +296,37 @@ MasterGraph::build()
+    auto *new_tensor = new rocalTensor(info);
+    // if the tensor is not an output tensor, the tensor creation is deferred and later it'll be created as a virtual tensor
+    if(is_output) {
+        if (new_tensor->create_from_handle(_context) != 0)
```
why do we need new_tensor and output?

fiona-gladwin replied:
In image pipeline the output is returned to user by getting read buffers from ring buffer and copying the data into user passed buffers. 
In the current changes we are maintaining 2 tensor lists. The internal_tensor_list which is used to store a list of the output tensors and used in output routine(handle swapped with the write buffers in ring buffer) . The output_tensor_list which stores the same list of output tensors and is used to return the output tensor list to the user (handle swapped with the read buffers in ring buffer when output list is requested by the user).  Hence we are creating the new_tensor and output to be pushed into respective tensor lists.

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -85,17 +90,31 @@ bool operator==(const rocalTensorInfo &rhs, const rocalTensorInfo &lhs) {
+    size_t roi_size = _batch_size * 4 * sizeof(unsigned);
+    if(_mem_type == RocalMemType::HIP) {
+#if ENABLE_HIP
+    hipError_t err = hipHostMalloc((void **)&_roi_buf, roi_size, hipHostMallocDefault/*hipHostMallocMapped|hipHostMallocWriteCombined*/);
```
remove the commented flags

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -85,17 +90,31 @@ bool operator==(const rocalTensorInfo &rhs, const rocalTensorInfo &lhs) {
+void rocalTensorInfo::allocate_tensor_roi_buffers() {
+    size_t roi_size = _batch_size * 4 * sizeof(unsigned);
+    if(_mem_type == RocalMemType::HIP) {
+#if ENABLE_HIP
```
add the condition for OPENCL

fiona-gladwin replied:
Since there is no OPENCL support for augmentations in RPP we have not added the allocation for OpenCL ROI buffers

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -124,6 +143,45 @@ rocalTensorInfo::rocalTensorInfo(std::vector<size_t> dims,
+rocalTensorInfo::~rocalTensorInfo() {
+    if(!_is_metadata) {
+        if(_mem_type == RocalMemType::HIP) {
+#if ENABLE_HIP
```
OPENCL is not handled here

fiona-gladwin replied:
We will create a separate PR for OpenCL changes

rrawther reviewed (changes requested):
@fiona-gladwin : please address the review comments. 
the tot changes needs to be merged. We need to build for CPU/OPENCL/HIP backends. If OPENCL is not supported for a node, it needs to throw error instead of crashing when running applications like imageaug

fiona-gladwin committed:
Rename the NUM_OF_DIMS preprocessor directive in OpenVX

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor3

fiona-gladwin committed:
Remove batchsize argument from Brightness and CMN Nodes

fiona-gladwin committed:
Rename variables to multiplier and offset in CMN

fiona-gladwin committed:
Remove enum conversion API for layout and datatype

For Brightness and CMN nodes

fiona-gladwin committed:
Add support for single handle creation in VX RPP

fiona-gladwin committed:
Fix CMN node

Allocate vx_tensor to pass crop coordinates
Add API to allocate pinned or HOST mem

fiona-gladwin committed:
Minor change

Rename variables

rrawther commented on amd_openvx/openvx/api/vx_api.cpp:
```diff
@@ -9836,6 +9836,26 @@ VX_API_ENTRY vx_status VX_API_CALL vxQueryTensor(vx_tensor tensor, vx_enum attri
 					}
 				}
 #endif
+				else if (tensor_data_type == VX_TYPE_UINT32)
```
If we are just returning the pointer for host memory, we can return it as (vx_uint8 *) for all the data types, so in the api there is no need for checking various conditions. The calling routine can typecast the pointer based on the data type. Also the (size == sizeof(vx_unit8 *)) is correct. Size is size of data pointer, not data_type

rrawther commented on amd_openvx_extensions/amd_rpp/source/Brightness.cpp:
```diff
@@ -21,96 +21,93 @@ THE SOFTWARE.
-        if (data->in_tensor_type == vx_type_e::VX_TYPE_UINT8 && data->out_tensor_type == vx_type_e::VX_TYPE_UINT8)
-        {
+    } else if (data->deviceType == AGO_TARGET_AFFINITY_CPU) {
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HOST, &data->roiTensorPtr, sizeof(vx_uint32)));
```
Please remove all the conditions for different data_types here and modify the vxQueryTensor to return one type of pointer similar to the call for VX_TENSOR_BUFFER_HIP

rrawther commented on amd_openvx_extensions/amd_rpp/source/Brightness.cpp:
```diff
@@ -21,96 +21,93 @@ THE SOFTWARE.
     vx_enum scalar_type;
     STATUS_ERROR_CHECK(vxQueryScalar((vx_scalar)parameters[5], VX_SCALAR_TYPE, &scalar_type, sizeof(scalar_type)));
-    if (scalar_type != VX_TYPE_UINT32)
+    if (scalar_type != VX_TYPE_INT32)
```
Can you please combine all the invalid cases into single one with OR condition

fiona-gladwin replied:
We need to query the parameters each time and check it's type if it fails we are throwing custom error message for each parameter should that be combined into a single statement?

rrawther commented on amd_openvx_extensions/amd_rpp/source/CropMirrorNormalize.cpp:
```diff
@@ -21,127 +21,111 @@ THE SOFTWARE.
-        {
+    else if (data->deviceType == AGO_TARGET_AFFINITY_CPU) {
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HOST, &data->roiTensorPtr, sizeof(vx_uint32)));
+        if (data->inputTensorType == vx_type_e::VX_TYPE_UINT8 && data->outputTensorType == vx_type_e::VX_TYPE_UINT8) {
```
please follow the suggestion for brightness

rrawther commented on amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:
```diff
@@ -1986,36 +1979,115 @@ vx_node createNode(vx_graph graph, vx_enum kernelEnum, vx_reference params[], vx
+        if (deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_OPENCL
+            STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_ATTRIBUTE_AMD_OPENCL_COMMAND_QUEUE, &handle->cmdq, sizeof(handle->cmdq)));
+            rppCreateWithStreamAndBatchSize(&data->rppHandle, handle->cmdq, batchSize);
```
I thought we are creating rpphandle only once per pipeline

fiona-gladwin replied:
Yes, the RPP handle is created once for the first node and for the subsequent nodes if the handle is already created we are just incrementing the pointer, The release is also done in a similar manner

fiona-gladwin committed:
Minor changes

rrawther commented on amd_openvx_extensions/amd_rpp/source/CropMirrorNormalize.cpp:
```diff
@@ -21,127 +21,111 @@ THE SOFTWARE.
-    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[8], 0, data->nbatchSize * 3, sizeof(vx_float32), data->mean, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
-    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[9], 0, data->nbatchSize * 3, sizeof(vx_float32), data->std_dev, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
-    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[10], 0, data->nbatchSize, sizeof(vx_uint32), data->mirror, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[3], 0, data->srcDescPtr->n * data->srcDescPtr->c, sizeof(vx_float32), data->multiplier, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
```
are we going to use pinned memory for these parameters as we discussed before?

fiona-gladwin replied:
Will work on changes required for ParameterVX and use pinned memory for the parameters. Will issue this as a separate PR in future

rrawther commented on rocAL/rocAL/source/augmentations/geometry_augmentations/node_crop_mirror_normalize.cpp:
```diff
@@ -26,59 +26,106 @@ THE SOFTWARE.
     }
-    _mean_array = vxCreateArray(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, _batch_size);
-    _std_dev_array = vxCreateArray(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, _batch_size);
+    _multiplier_vx_array = vxCreateArray(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, multiplier_offset_array_size);
```
I thought we have discussed to switch all the parameters to use tensor and pinned memory

fiona-gladwin replied:
Some of these params are RocalParameters, we need to convert them to rocalTensors as discussed, we will work on the changes

rrawther reviewed (changes requested):
please address review comments

fiona-gladwin committed:
Add changes to query tensor buffer HOST

Add changes to query without datatype
Make changes to Brightness, CMN, Copy and Nop nodes
Maintain single ptr for HOST and DEV buffers

LakshmiKumar23 commented:
@rrawther @fiona-gladwin is this ready to be merged?

fiona-gladwin commented:
> @rrawther @fiona-gladwin is this ready to be merged?

Hi @LakshmiKumar23 I am working on one particular comment. Once that is fixed I will report here.

fiona-gladwin commented:
Hi @rrawther and @LakshmiKumar23 I have addressed all the review comments in this PR

rrawther commented on rocAL/rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -297,33 +297,34 @@ MasterGraph::build()
         if (output->create_from_handle(_context) != 0)
-            THROW("Cannot create the image from handle")
+            THROW("Cannot create the tensor from handle")
+        _internal_tensor_list.push_back(output);
```
why do we need 2 sets of tensors (internal and output) for each tensor. This will increase the memory requirements twice

fiona-gladwin replied:
Hi @rrawther In image pipeline the output is returned to user by getting read buffers from ring buffer and copying the data into user passed buffers.
In the current changes we are maintaining 2 tensor lists in order to directly return the tensor list to the user. 

1. The internal_tensor_list which is used to store a list of the output tensors and used in output routine(handle swapped with the write buffers in ring buffer) . 
2. The output_tensor_list which stores the same list of output tensors and is used to return the tensor list to the user (handle swapped with the read buffers in ring buffer when output list is requested by the user).

Hence we are maintaining two tensor lists for this purpose. We were not able to use a single tensor list, because by the time user requests for the output the internal tensor list is updated with next batch of data
